### ISSUE TITLE
Consider using zig cc to make prebuilt binaries

### ISSUE THREAD

**Issue by creationix** [luvit/luvi#231]
https://andrewkelley.me/post/zig-cc-powerful-drop-in-replacement-gcc-clang.html

I think this might enable us to build all the prebuilt luvi binaries using a single linux host such as our CI servers.

**Comment by rphillips** [comment]
Looks extremely promising.

**Comment by squeek502** [comment]
Was thinking about this as well. Would be interesting to try.

**Comment by zhaozg** [comment]
+1

**Comment by squeek502** [comment]
Played around with it a bit. Was able to get luv to compile but the resulting LuaJIT fails to load the resulting luv.so.

```
git clone https://github.com/luvit/luv.git
cd luv
export CC="zig cc"
export ASM="zig cc"
make
./build/luajit tests/run.lua
PANIC: unprotected error in call to Lua API (error loading module 'luv' from file './luv.so':
        ./luv.so: undefined symbol: lua_status
stack traceback:
        [C]: at 0x002998b0
        [C]: in function 'require'
        ./lib/tap.lua:1: in main chunk
        [C]: in function 'require'
        tests/run.lua:6: in main chunk
        [C]: at 0x00218770)
```

`zig cc` is super new so I'd guess something from https://github.com/ziglang/zig/issues/4784 is the cause.

**Comment by andrewrk** [comment]
I'm happy to work with y'all on this, give me a couple days to fix up that flag integration like @squeek502 noted above and then let's coordinate on this use case.

For glibc versions, as long as you know which glibc version you want to target and we add some way to communicate that info to `zig cc` then it should work. I'm sure we can figure it out.

holy-build-box readme says:

> Binaries work on pretty much any glibc-2.12-or-later-based x86 and x86-64 Linux distribution released since approx 2011. A non-exhaustive list:

So you should be able to use `-target foo-bar-gnu.2.12`. That's zig-specific `-target` syntax, which is actually not supposed to work; I need to change it to be llvm syntax, but before making that change I'll make sure there's a way to specify the glibc version. It seems that two wrongs do make a right after all :)

**Comment by squeek502** [comment]
I have a feeling the tough part will be getting OpenSSL to compile. We [use a CMake `ExternalProject`](https://github.com/luvit/luvi/blob/master/deps/openssl.cmake) for it and call out to `make`/`nmake` and let OpenSSL do its (fairly complicated) thing. Our `tiny` builds (without OpenSSL and a few other things) will probably be pretty easy to get working by comparison.

**Comment by andrewrk** [comment]
One problem with cmake I learned this week is that it does not have a separate concept of a "host compiler" and a "target compiler". So unless the upstream project adds support for it, there will be an issue when a project builds a binary and then tries to run it. A better build system would allow you to choose which binaries were native and which were for the target, so that you could build your native tools natively, and build your target code for the target.

**Comment by creationix** [comment]
Instead of using -build-box gcc, I'd prefer if we could do a static build with musl.  Then it will work on even more linux versions than our old binaries.  I had trouble before trying to build static luajit binaries, but I seem to remember that was exactly one of the examples in @andrewrk's blog post.

**Comment by squeek502** [comment]
Was actually able to build a static musl binary of the luvi `tiny` variant by doing:

```
export CC="zig cc -target native-native-musl -fno-sanitize=undefined"
export ASM="zig cc"
make tiny
make
```
(the `-fno-sanitize=undefined` is to get around `Illegal instruction` errors from generating `vm_arch.h` during the LuaJIT build process, see [this comment](https://github.com/ziglang/zig/issues/4830#issuecomment-605491606)).

However, when running the resulting `luvi` binary, I'm getting:

```
PANIC: unprotected error in call to Lua API ([string "return require('init')(...)"]:1: module 'init' not found:
```

meaning that something is going wrong with the embedding of `src/lua/init.lua` via CMake [here](https://github.com/luvit/luvi/blob/8fd9c8b637b3a312c14dec7af5eae0e941f4fe5d/cmake/Modules/LuaJITAddExecutable.cmake#L18-L49).

**Comment by squeek502** [comment]
Oh, just remembered that I tried making a `musl` build of `luv` and ran into:

```
PANIC: unprotected error in call to Lua API (error loading module 'luv' from file './luv.so':
        Dynamic loading not supported
```

That is, when compiled statically, `musl` only has a [stub for dlopen](https://github.com/ziglang/zig/blob/master/lib/libc/musl/src/ldso/dlopen.c). This means that fully static builds are probably untenable for Luvi, since dynamically loading modules is definitely necessary.

**Comment by SinisterRectus** [comment]
Also https://github.com/phusion/holy-build-box/blob/master/README.md#why-statically-linking-to-glibc-is-a-bad-idea

**Comment by creationix** [comment]
> This means that fully static builds are probably untenable for Luvi, since dynamically loading modules is definitely necessary.

True, anybody needing to load native modules will need this.  I wonder if it's worth the effort to have a static build that disables loading native modules.  Even ffi use cases often use dlopen to call the libraries.

**Comment by andrewrk** [comment]
The other option zig cc provides you is "static except glibc" builds, and if you pick an old enough glibc version (e.g. matching holy-build-box) then your binary tarballs will work on any linux distro that uses glibc and standard dynamic linker path (which is most).

The nice thing about static tarballs is they work on *all* linux distros, even ones like NixOS which have non-standard dynamic linker paths, and alpine linux which uses musl instead of glibc.

**Comment by andrewrk** [comment]
Alright, I've done some polishing on this feature, and it's about 1 week until the 0.6.0 release. If you're still exploring this use case, now would be a good time to try it out and report issues that you run into. Let me know if I can help.

**Comment by squeek502** [comment]
Being able to target a specific glibc version would be nice. I'll look into the embedded-Lua file issue mentioned [here](https://github.com/luvit/luvi/issues/231#issuecomment-606503093) and try to figure out what's going wrong and how/if its related to `zig cc`.

**Comment by andrewrk** [comment]
I'm not planning to solve https://github.com/ziglang/zig/issues/4911 before the release, so you can use this syntax:

`-target x86_64-linux-gnu.2.12`

**Comment by squeek502** [comment]
`-target x86_64-linux-gnu.2.12` doesn't seem to affect the entire build process:

```
cd luvi
export CC="zig cc -target x86_64-linux-gnu.2.12 -fno-sanitize=undefined"
make tiny
make
```
results in:
```
lld: error: undefined symbol: fcntl64
>>> referenced by core.c
>>>               core.c.o:(uv__nonblock_fcntl) in archive deps/luv/deps/libuv/libuv_a.a
>>> referenced by core.c
>>>               core.c.o:(uv__nonblock_fcntl) in archive deps/luv/deps/libuv/libuv_a.a
>>> referenced by core.c
>>>               core.c.o:(uv__cloexec_fcntl) in archive deps/luv/deps/libuv/libuv_a.a
>>> referenced by core.c
>>>               core.c.o:(uv__cloexec_fcntl) in archive deps/luv/deps/libuv/libuv_a.a
>>> referenced by pipe.c
>>>               pipe.c.o:(uv_pipe_open) in archive deps/luv/deps/libuv/libuv_a.a
>>> referenced by process.c
>>>               process.c.o:(uv__process_child_init) in archive deps/luv/deps/libuv/libuv_a.a
>>> referenced by tty.c
>>>               tty.c.o:(uv_tty_init) in archive deps/luv/deps/libuv/libuv_a.a
>>> did you mean: fcntl64
>>> defined in: /home/ryan/.cache/zig/stage1/h/OPkcZeqDw6f_ToblL-yxpuotOgXOhqW3rELarafcu_hTHg9NvBniwDctKo9FnM-I/libc.so.6.0.0
```

**Comment by zhaozg** [comment]
I begin to learn "zig cc" 

build for macos
```
export CC="zig cc"
./Configure darwin64-x86_64-cc
make
Zig attempted to find the path to native system libc headers by executing this command:
zig cc -E -Wp,-v -xc /dev/null
Unable to link against libc: Unable to find libc installation: unable to spawn system C compiler
S
```
but cross build windows on macos success with
```
export CC="zig cc -target x86_64-windows-gnu"
export AR=x86_64-w64-mingw32-ar
./Configure no-shared mingw64
make
```
```
fd openssl.exe
apps/openssl.exe
apps/zig-cache/o/Jb6evqm5OTXTMdxGG9d_enVeByLdKBKqY7ynvUa8kdsu3lQ_pCCSjqVxKhbonni-/openssl.exe
```
great, @andrewrk, and I hope a `zig ar`

**Comment by truemedian** [comment]
It's been a while since this has been touched up on, but we can now obtain a run-able (but non-functional) luvi using zig.

Building on linux:
```shell
$ export CC="zig cc -target native-linux-gnu.2.28"
$ export CXX="zig c++ -target native-linux-gnu.2.28"
$ export ASM="zig cc -target native-linux-gnu.2.28"
$ make tiny
$ make
```

Luvi can be run:
```
$ ./build/luvi --version
./build/luvi v2.11.0-9-gc11154c
rex: 8.37 2015-04-28
libuv: 1.36.0
ssl: OpenSSL 1.1.1g  21 Apr 2020, lua-openssl 0.7.8
```

But the first test case (samples/test.app) panics with the following error:
```
$ ./build/luvi samples/test.app -- 1 2 3 4
PANIC: unprotected error in call to Lua API (7)
```

Which happens on (using lldb) the following traceback:
```
  * frame #0: 0x000000000027c760 luvi`panic
    frame #1: 0x0000000000264140 luvi`lj_err_throw + 112
    frame #2: 0x000000000029e028 luvi`lj_trace_err + 56
    frame #3: 0x000000000029d4b8 luvi`lj_record_ins + 18232
    frame #4: 0x0000000000292f6b luvi`trace_state + 1627
    frame #5: 0x00000000002f1b40 luvi`lj_vm_cpcall + 77
    frame #6: 0x000000000026c846 luvi`lj_dispatch_ins + 326
    frame #7: 0x00000000002f3500 luvi`lj_vm_inshook + 49
    frame #8: 0x0000000000275d2b luvi`lua_pcall + 155
    frame #9: 0x00000000002463e7 luvi`main + 215
    frame #10: 0x00007ffff7e1b152 libc.so.6`__libc_start_main + 242
    frame #11: 0x000000000023791a luvi`_start at start.S:120
```

**Comment by zhaozg** [comment]
I contine todo this with a checklist, after all build pass, I will make a PR to finish this job.

*toolchain: zig+cmake+make
*build machine: macos 11

- [ ] zig cc --target x86_64-macos-gnu
- [ ] zig cc --target x86_64-windows-gnu
- [ ] zig cc --target x86_64-linux-gnu
- [ ] zig cc --target x86_64-aarch64-gnu

**Comment by andrewrk** [comment]
Good news- thanks to @kubkon's [efforts](https://twitter.com/kubkon/status/1333506865683116035), you can soon add `aarch64-macos-gnu` to the list :)

**Comment by zhaozg** [comment]
I have build luvi success, but fail to run it.
run `lldb -- luvi $LUVI_SRC/sample/test.app`
got 
```
{ }
{ args = { [0] = '/Users/zhaozg/work/build/luvit/native/luvi/luvi' },
  bundle = { mainPath = 'main.lua', readfile = function: 0x0100789610,
    readdir = function: 0x0100789560, register = function: 0x01007873a0,
    base = '/Users/zhaozg/work/luvit/luvi/samples/test.app',
    paths = { '/Users/zhaozg/work/luvit/luvi/samples/test.app' }, uvi was compiled with optimization - stepping may behave oddly; variables may not be available.
Process 66737 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 1.18
    frame #0: 0x00000001001b82f4 luvi`panic(L=0x0000000100763380) at lib_aux.c:313:19 [opt]
   310
   311  static int panic(lua_State *L)
   312  {
-> 313    const char *s = lua_tostring(L, -1);
   314    fputs("PANIC: unprotected error in call to Lua API (", stderr);
   315    fputs(s ? s : "?", stderr);
   316    fputc(')', stderr); fputc('\n', stderr);
Target 0: (luvi) stopped.
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 1.18
  * frame #0: 0x00000001001b82f4 luvi`panic(L=0x0000000100763380) at lib_aux.c:313:19 [opt]
(lldb)
```
I have no ideas, and just guest not compat with JIT feature

**Comment by truemedian** [comment]
My latest attempt at building luvi.

TLDR; I now have luvi building for x86_64-linux-gnu on linux.

# Successes

- [x] `x86_64-linux-gnu.2.28` (tiny): working luvi, passes all tests
- [x] `x86_64-linux-gnu.2.28` (regular): working luvi, passes all tests

# Bulding

Using zig version: `0.8.0-dev.945+166e9ea68`

## Setup

zig-cc: `zig cc -target $ZIG_TARGET $@`

zig-asm: `zig cc -target $ZIG_TARGET $@`

zig-cxx: `zig c++ -target $ZIG_TARGET $@`

```sh
export CC="$(pwd)/zig-cc"
export CXX="$(pwd)/zig-cxx"
export ASM="$(pwd)/zig-asm"
```

CMAKE_FLAGS: `-H. -Bbuild -DCMAKE_BUILD_TYPE=Release -DWithSharedLibluv=OFF -DCMAKE_C_COMPILER=$CC -DCMAKE_CXX_COMPILER=$CXX -DCMAKE_ASM_COMPILER=$ASM`

```sh
make tiny "CMAKE_FLAGS=$CMAKE_FLAGS"
make
make test
```

## Targets

### Much Bug

- [ ] `mips64el-linux-gnuabi64`: zig: unable to build C object: FileNotFound

- [ ] `mips64el-linux-gnuabin32`: ldd: testCCompiler.o incompatible with elf64ltsmip

- [ ] `mips64-linux-gnuabi64`: zig: unable to build C object: FileNotFound

- [ ] `mips64-linux-gnuabin32`: ldd: testCCompiler.o incompatible with elf64btsmip

- [ ] `mipsel-linux-gnu`: zig: unable to build C object: FileNotFound

- [ ] `mips-linux-gnu`: zig: unable to build C object: FileNotFound

- [ ] `powerpc64le-linux-gnu`: 'sysdeps/unix/sysv/linux/powerpc/sysdep.h' not found

- [ ] `powerpc64-linux-gnu`: 'sysdeps/unix/sysv/linux/powerpc/sysdep.h' not found

- [ ] `powerpc-linux-gnu`: zig: unknown target CPU 'ppc32'

- [ ] `x86_64-linux-gnux32`: libunwind: UnwindCursor<> does not fit in unw_cursor_t

### Less Bug

Some of these are probably caused by a lack of system support (see emulation)

- [ ] `aarch64_be-linux-gnu`: ldd: unknown emulation: aarch64_be_linux

- [ ] `aarch64_be-windows-gnu`: zig mingw: only win32 supported

- [ ] `aarch64-linux-gnu`: dlopen not found

- [ ] `aarch64-windows-gnu`: '../bsd_private_base.h' not found

- [ ] `armeb-linux-gnueabi`: ldd: unknown emulation: armebelf_linux_eabi

- [ ] `armeb-linux-gnueabihf`: ldd: unknown emulation: armebelf_linux_eabi

- [ ] `armeb-windows-gnu`: cmake: backend data layout mismatch target description

- [ ] `arm-linux-gnueabi`: 'arm-features.h' not found

- [ ] `arm-linux-gnueabihf`: 'arm-features.h' not found

- [ ] `arm-windows-gnu`: cmake: backend data layout mismatch target description

- [ ] `i386-linux-gnu`: ldd: undefined hidden symbol: __x86.get_pc_thunk.bx

- [ ] `i386-windows-gnu`: dlopen not found

- [x] `x86_64-linux-gnu`: versioned fnctl64: use gnu.2.28

- [ ] `x86_64-windows-gnu`: dlopen not found

- [ ] `x86_64-macos-gnu`: dlopen not found

The biggest issue for major systems right now is linking dlopen

**Comment by creationix** [comment]
Would a version that disables dlopen be useful?  It would break loading native modules as well as dynamic linking to libraries using ffi, but we would still have access to all the built-in native code and can still use ffi for ctypes type stuff.

**Comment by truemedian** [comment]
Seen above, @zhaozg did get cross compiling to work for most of the systems we care about, but I assume thats after modifying the build system to make proper assumptions about cross compilation. Not linking dlopen is actually an issue because of how luvi loads itself, its compiled as a dynamic library which loads itself, which is how the precompiled luajit stuff works (init, luvibundle, luvipath are all loaded this way). If we were to link without dlopen, those would no longer work and everything would break.

**Comment by creationix** [comment]
Bleh, I forgot about that part.  At least we're making good progress.

**Comment by zhaozg** [comment]
> Seen above, @zhaozg did get cross compiling to work for most of the systems we care about.

That list is obsolete, some build broken or output run crash, I need recheck it, and we should give more time to wait zig grows
